airbyte-cdk/python/airbyte_cdk/sources/file_based/file_based_stream.py

```diff
@@ -0,0 +1,163 @@
+        """
+        if self.config.input_schema:
+            return type_mapping_to_jsonschema(self.config.input_schema)
+
+        files = self.list_files()
+        if len(files) > MAX_N_FILES_FOR_STREAM_SCHEMA_INFERENCE:
+            files = sorted(files, key=lambda x: x.last_modified, reverse=True)[:MAX_N_FILES_FOR_STREAM_SCHEMA_INFERENCE]
+            logging.warning(f"Refusing to infer schema for {len(files)} files; using {MAX_N_FILES_FOR_STREAM_SCHEMA_INFERENCE} files.")
+        return self.infer_schema(files)
+
+    @cache
+    def list_files(self) -> List[RemoteFile]:
+        """
+        List all files that belong to the stream as defined by the stream's
+        globs.
+        """
+        return list(self.stream_reader.list_matching_files(self.config.globs))
+
+    def list_files_for_this_sync(self, stream_state: Mapping[str, Any]) -> Iterable[RemoteFile]:
+        """
+        Return the subset of this stream's files that will be read in the current sync.
+
+        Specifically:
+
+        - Take the output of `list_files`
+        - If `stream_state.start` is non-None then remove all files with last_modified
+          date before `start`
+        """
+        # TODO: implement me
+        return iter(self.list_files())
+
+    def infer_schema(self, files: List[RemoteFile]) -> Mapping[str, Any]:
+        loop = asyncio.get_event_loop()
+        return loop.run_until_complete(self._infer_schema(files))
+
+    async def _infer_schema(self, files: List[RemoteFile]) -> Mapping[str, Any]:
+        """
+        Infer the schema for a stream.
+
+        Each file type has a corresponding `infer_schema` handler.
+        Dispatch on file type.
+        """
+        base_schema = {}
+        pending_tasks = set()
+
+        n_started, n_files = 0, len(files)
+        files = iter(files)
+        while pending_tasks or n_started < n_files:
+            while len(pending_tasks) < self.discovery_concurrency_policy.n_concurrent_requests and (file := next(files, None)):
+                pending_tasks.add(asyncio.create_task(self._infer_file_schema(file)))
+                n_started += 1
+            # Return when the first task is completed so that we can enqueue a new task as soon as the
+            # number of concurrent tasks drops below the number allowed.
+            done, pending_tasks = await asyncio.wait(pending_tasks, return_when=asyncio.FIRST_COMPLETED)
+            for task in done:
+                base_schema = merge_schemas(base_schema, task.result())
+
+        return base_schema
+
+    async def _infer_file_schema(self, file: RemoteFile) -> Mapping[str, Any]:
+        try:
+            return await self._get_parser(self.config.file_type).infer_schema(file, self.stream_reader)
+        except Exception as exc:
+            raise SchemaInferenceError(f"Error inferring schema for file: {file.uri}") from exc
+
+    @staticmethod
+    def _get_parser(file_type: FileType) -> FileTypeParser:
```

Comment:
is it crazy to say that this shouldn't be static/private, and that a user implementing a file-based connector should be able to supply a parser to support a new filetype in their connector without having to upstream it to the CDK? 

if this is a lot of work, we should do it as an additive change. But if it's easy, might as well do it out the gate. It seems to me like this particular function should be extensible by the user

Reply:
This shouldn't be a lot of work. I'll look into this as part of a reorganization to give us more stream-level dependency injection.